state/in-memory/in_memory.go

```diff
@@ -0,0 +1,389 @@
+	return []state.Feature{state.FeatureETag, state.FeatureTransactional}
+}
+
+func (store *inMemoryStore) Delete(req *state.DeleteRequest) error {
+	// step1: validate parameters
+	if err := store.doDeleteValidateParameters(req); err != nil {
+		return err
+	}
+
+	// step2 and step3 should be protected by write-lock
+	store.lock.Lock()
+	defer store.lock.Unlock()
+
+	// step2: validate etag if needed
+	if err := store.doValidateEtag(req.Key, req.ETag, req.Options.Concurrency); err != nil {
+		return err
+	}
+
+	// step3: do really delete
+	// this operation won't fail
+	store.doDelete(req.Key)
+	return nil
+}
+
+func (store *inMemoryStore) doDeleteValidateParameters(req *state.DeleteRequest) error {
+	return state.CheckRequestOptions(req.Options)
+}
+
+func (store *inMemoryStore) doValidateEtag(key string, etag *string, concurrency string) error {
+	if etag != nil && *etag != "" && concurrency == state.FirstWrite {
+		// For FirstWrite, we need to validate etag before delete
+		item := store.items[key]
+		if item == nil {
+			return state.NewETagError(state.ETagMismatch, fmt.Errorf("state not exist or expired for key=%s", key))
+		}
+		if item.etag != etag {
+			return state.NewETagError(state.ETagMismatch, fmt.Errorf(
+				"state etag not match for key=%s: current=%s, expect=%s", key, *item.etag, *etag))
+		}
+	}
+	return nil
+}
+
+func (store *inMemoryStore) doDelete(key string) {
+	delete(store.items, key)
+}
+
+func (store *inMemoryStore) BulkDelete(req []state.DeleteRequest) error {
+	if len(req) == 0 {
+		return nil
+	}
+
+	// step1: validate parameters
+
+	for i := 0; i < len(req); i++ {
+		if err := store.doDeleteValidateParameters(&req[i]); err != nil {
+			return err
+		}
+	}
+
+	// step2 and step3 should be protected by write-lock
+	store.lock.Lock()
+	defer store.lock.Unlock()
+
+	// step2: validate etag if needed
+	for _, dr := range req {
+		err := store.doValidateEtag(dr.Key, dr.ETag, dr.Options.Concurrency)
+		if err != nil {
+			return err
+		}
+	}
+
+	// step3: do really delete
+	for _, dr := range req {
+		store.doDelete(dr.Key)
+	}
+	return nil
+}
+
+func (store *inMemoryStore) Get(req *state.GetRequest) (*state.GetResponse, error) {
+	item := store.doGetWithReadLock(req.Key)
+	if item != nil && isExpired(item.expire) {
+		store.doDeleteExpiredWithWriteLock(req.Key)
+		item = nil
+	}
+
+	if item == nil {
+		return &state.GetResponse{Data: nil, ETag: nil}, nil
+	}
+	return &state.GetResponse{Data: unmarshal(item.data), ETag: item.etag}, nil
+}
+
+func (store *inMemoryStore) doGetWithReadLock(key string) *inMemStateStoreItem {
+	store.lock.RLock()
+	defer store.lock.RUnlock()
+
+	return store.items[key]
+}
+
+func (store *inMemoryStore) doDeleteExpiredWithWriteLock(key string) {
+	store.lock.Lock()
+	defer store.lock.Unlock()
+	// get item and check expired again to avoid if item changed between we got this write-lock
+	item := store.items[key]
+	if item != nil && isExpired(item.expire) {
+		store.doDelete(key)
+	}
+}
```

Comment:
This func should return the item and a bool value to indicate wether it is deleted(or just the item, null for deleted). If the item is not deleted, `Get` should returns the item instead of nil.

Reply:
fixed